File: Java/src/Palindrome Partitioning.java
public class Solution {
    public List<List<String>> partition(String s) {
        List<List<String>> list = new ArrayList<>();
        if (s==null || s.length() == 0) return list;
        return partition(s, 0, s.length());
    }

    public List<List<String>> partition(String s, int start, int end) {
        List<List<String>> list = new ArrayList<>();

        if (isPalidrome(s,start,end)){
            List<String> l = new ArrayList<>();
            l.add(s.substring(start, end));
            list.add(l);
        }

        if (end-start == 1) return list;

        for (int i = start+1; i<end; i++){
            if (isPalidrome(s,start,i)){
                List<List<String>> newList = partition(s, i, end);
                for(List<String> j : newList){
                    j.add(0,s.substring(start,i));
                }
                list.addAll(newList);
            }
        }
        return list;
    }

    private boolean isPalidrome(String s, int start, int end) {
        for (int i=start, j=end-1; j>i; j--,i++) {
            if (s.charAt(i)!=s.charAt(j))
                return false;
        }

        return true;
    }
}

File: Python/Palindrome Partitioning.py
class Solution:
    # @param s, a string
    # @return a list of lists of string
    def partition(self, s):
        if not s:
            return [[]]
        return self.partitionHelper(s, 0 ,len(s))
    
    def partitionHelper( self, s, start, end):
        ret = []
        if self.isPalindrome(s[start:end]):
            ret.append([s[start:end]])
        if end - start == 1:
            return ret
        for i in range(start+1, end):
            if self.isPalindrome(s[start:i]):
                newList = self.partitionHelper(s, i, end)
                for l in newList:
                    print newList,l
                    l.insert(0, s[start:i])
                ret.extend(newList)
        return ret
        
        
    def isPalindrome(self,s):
        return s[::-1] == s
    
s = Solution()
print s.partition("ab")